applications/CompressiblePotentialFlowApplication/custom_processes/apply_far_field_process.cpp

```diff
@@ -0,0 +1,129 @@
+        if (min_projections[i_thread] < min_projections[minimum_node_thread_id]){
+            minimum_node_thread_id = i_thread;
+        }
+    }
+
+    // Storing final result in member variable
+    mpReferenceNode = mrModelPart.pGetNode(nodes_id_list[minimum_node_thread_id]);
+}
+
+void ApplyFarFieldProcess::AssignFarFieldBoundaryConditions()
+{
+    double velocity_projection = 0.0;
+    array_1d<double,3> normal;
+    array_1d<double,3> aux_coordinates;
+
+    #pragma omp parallel for firstprivate(velocity_projection, normal, aux_coordinates)
+    for (int i = 0; i < static_cast<int>(mrModelPart.Conditions().size()); i++) {
+        auto it_cond = mrModelPart.ConditionsBegin() + i;
+        auto& r_geometry = it_cond->GetGeometry();
+        // Computing normal
+        r_geometry.PointLocalCoordinates(aux_coordinates, r_geometry.Center());
+        normal = it_cond->GetGeometry().Normal(aux_coordinates);
+
+        velocity_projection = inner_prod(normal, mFreeStreamVelocity);
+        if (velocity_projection < 0.0) {
+            AssignDirichletFarFieldBoundaryCondition(r_geometry);
+        }
+        else {
+            AssignNeumannFarFieldBoundaryCondition(*it_cond);
+        }
+
+    }
+}
+
+void ApplyFarFieldProcess::AssignDirichletFarFieldBoundaryCondition(Geometry<NodeType>& rGeometry)
+{
+    // Fixing nodes in the domain that are part of the inlet, and initializing its value
+    // according to the free stream velocity.
+    for (std::size_t i_node = 0; i_node < rGeometry.size(); i_node++){
+        array_1d<double,3> relative_coordinates = rGeometry[i_node].Coordinates() - mpReferenceNode->Coordinates();
+        double inlet_potential = inner_prod(relative_coordinates, mFreeStreamVelocity);
+        rGeometry[i_node].Fix(VELOCITY_POTENTIAL);
+        rGeometry[i_node].FastGetSolutionStepValue(VELOCITY_POTENTIAL) = inlet_potential + mReferencePotential;
+        // Checking if its an adjoint case.
+        if (rGeometry[i_node].SolutionStepsDataHas(ADJOINT_VELOCITY_POTENTIAL)){
+            rGeometry[i_node].Fix(ADJOINT_VELOCITY_POTENTIAL);
+            rGeometry[i_node].FastGetSolutionStepValue(ADJOINT_VELOCITY_POTENTIAL) = 0.0;
+        }
+    }
+}
+
+void ApplyFarFieldProcess::AssignNeumannFarFieldBoundaryCondition(Condition& rCondition)
+{
+    rCondition.SetValue(FREE_STREAM_VELOCITY, mFreeStreamVelocity);
+}
+
+void ApplyFarFieldProcess::InitializeFlowField()
+{
+    ModelPart& root_model_part = mrModelPart.GetRootModelPart();
+    double inlet_potential = 0.0;
+    array_1d<double,3> relative_coordinates;
+
+    #pragma omp parallel for firstprivate(inlet_potential,relative_coordinates)
+    for (int i = 0; i < static_cast<int>(root_model_part.Nodes().size()); i++) {
+        auto it_node = root_model_part.NodesBegin() + i;
+
+        relative_coordinates = it_node->Coordinates() - mpReferenceNode->Coordinates();
+        inlet_potential = inner_prod(relative_coordinates, mFreeStreamVelocity);
```

Comment:
I think initial_potential is more appropiate here

```suggestion
        const double initial_potential = inner_prod(relative_coordinates, mFreeStreamVelocity);
```